Implement percentage values for letter-spacing and word-spacing per css-text-4

fantasai opened the pull request:
#### 6abbfb00701a455534e7d9af010909a6dc2b66ed
<pre>
Implement percentage values for letter-spacing and word-spacing per css-text-4
<a href="https://bugs.webkit.org/show_bug.cgi?id=262762">https://bugs.webkit.org/show_bug.cgi?id=262762</a>
<a href="https://rdar.apple.com/problem/116562262">rdar://problem/116562262</a>

Reviewed by Antti Koivisto and Matthieu Dubet.

Updates property grammars to accept &lt;length-percentage&gt;.
Shifts storage of word-spacing and letter-spacing on FontCascade
  from float to Length so that it inherits correctly.
Updates float accessors to compute the Length against current font-size.
Adds computed value accessors.
Adapts FontCascade constructors to more cleanly support cloning vs resetting.

Note: letter-spacing and word-spacing values need to live on FontCascade
because they&apos;re used for text measurement routines, which don&apos;t have
access to RenderStyle.

* LayoutTests/css1/text_properties/letter_spacing.html:
* LayoutTests/css2.1/t1604-c542-letter-sp-01-b-a.html:
* LayoutTests/fast/css3-text/css3-word-spacing-percentage/word-spacing-percentage-expected.html: Removed.
* LayoutTests/fast/css3-text/css3-word-spacing-percentage/word-spacing-percentage.html: Removed.
* LayoutTests/fast/text/synthetic-bold-percentage-word-spacing-expected.html: Removed.
* LayoutTests/fast/text/synthetic-bold-percentage-word-spacing.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/inheritance.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-percent-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-percent-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/letter-spacing-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/letter-spacing-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/letter-spacing-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/letter-spacing-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/letter-spacing-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/letter-spacing-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-spacing-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-spacing-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-spacing-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-spacing-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-spacing-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-spacing-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-computed-001-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-computed-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-negative-value-001-expected.xht: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-negative-value-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-percent-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-percent-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/letter-spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/letter-spacing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/animations/calc-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/calc-letter-spacing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/minmax-length-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/minmax-length-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/platform/glib/css1/text_properties/letter_spacing-expected.txt:
* LayoutTests/platform/gtk/css1/text_properties/letter_spacing-expected.png:
* LayoutTests/platform/mac/css1/text_properties/letter_spacing-expected.png:
* LayoutTests/platform/mac/css1/text_properties/letter_spacing-expected.txt:
* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::InspectorOverlay::drawRulers):
* Source/WebCore/inspector/InspectorOverlayLabel.cpp:
(WebCore::systemFont):
* Source/WebCore/layout/formattingContexts/inline/InlineLine.h:
* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::NonFastScrollableRegionOverlay::drawRect):
(WebCore::InteractionRegionOverlay::drawSettings):
* Source/WebCore/page/linux/ResourceUsageOverlayLinux.cpp:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::FontCascade):
(WebCore::FontCascade::operator=):
(WebCore::FontCascade::operator== const):
(WebCore::FontCascade::letterSpacing const):
(WebCore::FontCascade::wordSpacing const):
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::computedLetterSpacing const):
(WebCore::FontCascade::computedWordSpacing const):
(WebCore::FontCascade::setLetterSpacing):
(WebCore::FontCascade::setWordSpacing):
(WebCore::FontCascade::textAutospace const):
(WebCore::FontCascade::equalForTextAutoSizing const):
(WebCore::FontCascade::Spacing::Spacing):
(WebCore::FontCascade::Spacing::operator== const):
(WebCore::FontCascade::wordSpacing const): Deleted.
(WebCore::FontCascade::letterSpacing const): Deleted.
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::drawText):
* Source/WebCore/platform/win/DragImageWin.cpp:
(WebCore::dragLabelFont):
* Source/WebCore/rendering/RenderCombineText.cpp:
(WebCore::RenderCombineText::combineTextIfNeeded):
* Source/WebCore/rendering/RenderEmbeddedObject.cpp:
(WebCore::RenderEmbeddedObject::getReplacementTextGeometry const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::patternForDescription):
* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::bolder):
(WebCore::RenderListBox::paintItemForeground):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::computedWordSpacing const):
(WebCore::RenderStyle::computedLetterSpacing const):
(WebCore::RenderStyle::setFontDescription):
(WebCore::RenderStyle::setLetterSpacing):
(WebCore::RenderStyle::setWordSpacing):
(WebCore::RenderStyle::wordSpacing const): Deleted.
(WebCore::RenderStyle::letterSpacing const): Deleted.
(WebCore::RenderStyle::setLetterSpacingWithoutUpdatingFontDescription): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::initialLetterSpacing): Deleted.
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialLetterSpacing):
(WebCore::RenderStyle::letterSpacing const):
(WebCore::RenderStyle::wordSpacing const):
* Source/WebCore/rendering/svg/RenderSVGInlineText.cpp:
(WebCore::RenderSVGInlineText::computeNewScaledFontForStyle):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextLengthOrNormal):
(WebCore::Style::BuilderConverter::convertSpacing): Deleted.
(WebCore::Style::BuilderConverter::convertWordSpacing): Deleted.
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInheritWordSpacing):
(WebCore::Style::BuilderCustom::applyInheritLetterSpacing):
(WebCore::Style::BuilderCustom::applyInitialLetterSpacing):
(WebCore::Style::BuilderCustom::applyValueLetterSpacing):
(WebCore::Style::applyLetterSpacing): Deleted.

Canonical link: <a href="https://commits.webkit.org/271357@main">https://commits.webkit.org/271357@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80124c9e1a849909c1724ce1001e0e8a3bd05848

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28126 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6767 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30655 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/25635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28623 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8786 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4151 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/25622 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28394 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5528 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24199 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4781 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/4941 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25193 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31344 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25735 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25624 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/31250 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/4928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3098 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29007 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6473 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/24992 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6741 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5432 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -3443,10 +3443,15 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
+    case CSSPropertyLetterSpacing: {
+        const Length& spacing = style.computedLetterSpacing();
+        if (spacing.isFixed()) {
+            if (spacing.isZero())
+                return CSSPrimitiveValue::create(CSSValueNormal);
```
The computed value is always a length according to the spec: https://drafts.csswg.org/css-text/#letter-spacing-property

I would just remove the special case.

fantasai replied:
The computed value is, but the resolved value isn't.

>  For legacy reasons, a computed letter-spacing of zero yields a resolved value (getComputedStyle() return value) of normal.

nt1m replied:
I got confused by the table: "Computed value: | an absolute length" . Perhaps the spec should refer to the note in the table. Thanks for pointing it out!

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -3443,10 +3443,15 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
+                return CSSPrimitiveValue::create(CSSValueNormal);
+            return zoomAdjustedPixelValue(spacing.value(), style);
+        }
+        return CSSPrimitiveValue::create(spacing, style);
```
The spec says it should always compute to an absolute length. Is it what we're doing? Do we have tests that cover this? 

It sounds to me that what we want is to return the a pixel length containing the raw float value

fantasai replied:
No, the percentages need to be preserved in the computed value, otherwise they don't inherit correctly.

> Computed value: absolute length and/or percentage

And since there's no legacy compat constraints on percentages here, `getComputedStyle()` doesn't have a special exception (like `width` does) to return the used value instead of the computed value.

nt1m replied:
Ah I was reading CSS text 3 instead of 4, sorry.

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -3773,10 +3778,10 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
-            return zoomAdjustedPixelValue(style.fontCascade().wordSpacing(), style);
-        return CSSPrimitiveValue::create(wordSpacingLength, style);
+        const Length& spacing = style.computedWordSpacing();
+        if (spacing.isFixed())
+            return zoomAdjustedPixelValue(spacing.value(), style);
+        return CSSPrimitiveValue::create(spacing, style);
```
Same comment here re: absolute length.

nt1m commented on Source/WebCore/platform/graphics/FontCascade.h:
```diff
@@ -319,10 +321,16 @@ class FontCascade : public CanMakeWeakPtr<FontCascade> {
+        Length letter;
+        Length word;
+        Spacing() : letter(LengthType::Fixed) , word(LengthType::Fixed) { };
+        bool operator==(const Spacing& other) const { return letter == other.letter && word == other.word; }
```
The default operator should give you this.

fantasai replied:
That's what I thought, but it wouldn't compile without it. :/ Idk what to do.

nt1m commented on Source/WebCore/platform/graphics/FontCascade.h:
```diff
@@ -150,12 +151,14 @@ class FontCascade : public CanMakeWeakPtr<FontCascade> {
-    float letterSpacing() const { return m_letterSpacing; }
+    float letterSpacing() const;
+    float wordSpacing() const;
+    const Length& computedLetterSpacing() const { return m_spacing.letter; }
+    const Length& computedWordSpacing() const { return m_spacing.word; }
```
I would expect absolute pixel lengths with the name `computedLetterSpacing`, which I don't think we're necessarily returning here (it sounds like we might preserve percentages in here).

mdubet commented on Source/WebCore/layout/formattingContexts/inline/InlineLine.h:
```diff
@@ -28,6 +28,7 @@
 #include "InlineDisplayBox.h"
 #include "InlineItem.h"
 #include "InlineTextItem.h"
+#include "RenderStyleInlines.h"
```
Is this include needed ?

fantasai replied:
Yes, because RenderStyle::letterSpacing moved into RenderStyleInlines.h (from RenderStyle.cpp), and that means this line at the bottom of the page won't compile:
https://searchfox.org/wubkat/source/Source/WebCore/layout/formattingContexts/inline/InlineLine.h#370

mdubet commented on Source/WebCore/platform/graphics/FontCascade.h:
```diff
@@ -151,12 +152,14 @@ class FontCascade : public CanMakeWeakPtr<FontCascade>, public CanMakeCheckedPtr
+    float wordSpacing() const;
+    const Length& computedLetterSpacing() const { return m_spacing.letter; }
+    const Length& computedWordSpacing() const { return m_spacing.word; }
+    void setLetterSpacing(const Length& s) { m_spacing.letter = s; }
```
We generally don't use single letter parameter, maybe `length` ?

fantasai replied:
Sure, I can change it to `spacing`. The original was using `s` (see below), so I just left it unchanged.

mdubet commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2828,40 +2828,31 @@ WhiteSpace RenderStyle::whiteSpace() const
+        auto description = fontDescription();
+        description.setShouldDisableLigaturesForSpacing(fontCascade().letterSpacing());
+        setFontDescription(WTFMove(description));
+        fontCascade().update(selector);
```
Can we move the selector into the `update` function?

fantasai replied:
I was trying not to change the existing call pattern (see below, the diff is weird, maybe I should fix that).
Note that we're consistent about it throughout the codebase: https://searchfox.org/wubkat/search?q=FontCascade%3A%3Aupdate&path=&case=false&regexp=false

mdubet reviewed (approved):
🚀

anttijk commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2825,40 +2825,31 @@ WhiteSpace RenderStyle::whiteSpace() const
+    m_inheritedData.access().fontCascade.setLetterSpacing(WTFMove(spacing));
+
+    // Switching letter-spacing between zero and non-zero requires updating fonts (to enable/disable ligatures)
+    if (oldShouldDisableLigatures != bool(fontCascade().letterSpacing())) {
```
We usually prefer C++ style casts (static_cast<>) or conversion operators (!! for bool conversion). Here using a helper local (`auto shouldDisableLigatures == !!fontCascade().letterSpacing();`) would be more readable.

anttijk commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2825,40 +2825,31 @@ WhiteSpace RenderStyle::whiteSpace() const
+
+    // Switching letter-spacing between zero and non-zero requires updating fonts (to enable/disable ligatures)
+    if (oldShouldDisableLigatures != bool(fontCascade().letterSpacing())) {
+        auto selector = fontCascade().fontSelector();
```
It is good to document raw pointers with `auto*`.

anttijk reviewed (approved):
Looks fine. It is bit unclear to me why the spacing properties are now part of the FontCascade rather than being in RenderStyle directly (though I can't say it is wrong either). Maybe you could at least include some rationale to the commit message?

fantasai committed:
Implement percentage values for letter-spacing and word-spacing per css-text-4
https://bugs.webkit.org/show_bug.cgi?id=262762
rdar://problem/116562262

Reviewed by Antti Koivisto and Matthieu Dubet.

Updates property grammars to accept <length-percentage>.
Shifts storage of word-spacing and letter-spacing on FontCascade
  from float to Length so that it inherits correctly.
Updates float accessors to compute the Length against current font-size.
Adds computed value accessors.
Adapts FontCascade constructors to more cleanly support cloning vs resetting.

Note: letter-spacing and word-spacing values need to live on FontCascade
because they're used for text measurement routines, which don't have
access to RenderStyle.

* LayoutTests/css1/text_properties/letter_spacing.html:
* LayoutTests/css2.1/t1604-c542-letter-sp-01-b-a.html:
* LayoutTests/fast/css3-text/css3-word-spacing-percentage/word-spacing-percentage-expected.html: Removed.
* LayoutTests/fast/css3-text/css3-word-spacing-percentage/word-spacing-percentage.html: Removed.
* LayoutTests/fast/text/synthetic-bold-percentage-word-spacing-expected.html: Removed.
* LayoutTests/fast/text/synthetic-bold-percentage-word-spacing.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/inheritance.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-percent-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-percent-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/letter-spacing-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/letter-spacing-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/letter-spacing-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/letter-spacing-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/letter-spacing-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/letter-spacing-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-spacing-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-spacing-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-spacing-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-spacing-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-spacing-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-spacing-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-computed-001-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-computed-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-negative-value-001-expected.xht: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-negative-value-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-percent-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-percent-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/letter-spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/letter-spacing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/animations/calc-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/calc-letter-spacing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/minmax-length-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/minmax-length-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/platform/glib/css1/text_properties/letter_spacing-expected.txt:
* LayoutTests/platform/gtk/css1/text_properties/letter_spacing-expected.png:
* LayoutTests/platform/mac/css1/text_properties/letter_spacing-expected.png:
* LayoutTests/platform/mac/css1/text_properties/letter_spacing-expected.txt:
* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::InspectorOverlay::drawRulers):
* Source/WebCore/inspector/InspectorOverlayLabel.cpp:
(WebCore::systemFont):
* Source/WebCore/layout/formattingContexts/inline/InlineLine.h:
* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::NonFastScrollableRegionOverlay::drawRect):
(WebCore::InteractionRegionOverlay::drawSettings):
* Source/WebCore/page/linux/ResourceUsageOverlayLinux.cpp:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::FontCascade):
(WebCore::FontCascade::operator=):
(WebCore::FontCascade::operator== const):
(WebCore::FontCascade::letterSpacing const):
(WebCore::FontCascade::wordSpacing const):
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::computedLetterSpacing const):
(WebCore::FontCascade::computedWordSpacing const):
(WebCore::FontCascade::setLetterSpacing):
(WebCore::FontCascade::setWordSpacing):
(WebCore::FontCascade::textAutospace const):
(WebCore::FontCascade::equalForTextAutoSizing const):
(WebCore::FontCascade::Spacing::Spacing):
(WebCore::FontCascade::Spacing::operator== const):
(WebCore::FontCascade::wordSpacing const): Deleted.
(WebCore::FontCascade::letterSpacing const): Deleted.
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::drawText):
* Source/WebCore/platform/win/DragImageWin.cpp:
(WebCore::dragLabelFont):
* Source/WebCore/rendering/RenderCombineText.cpp:
(WebCore::RenderCombineText::combineTextIfNeeded):
* Source/WebCore/rendering/RenderEmbeddedObject.cpp:
(WebCore::RenderEmbeddedObject::getReplacementTextGeometry const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::patternForDescription):
* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::bolder):
(WebCore::RenderListBox::paintItemForeground):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::computedWordSpacing const):
(WebCore::RenderStyle::computedLetterSpacing const):
(WebCore::RenderStyle::setFontDescription):
(WebCore::RenderStyle::setLetterSpacing):
(WebCore::RenderStyle::setWordSpacing):
(WebCore::RenderStyle::wordSpacing const): Deleted.
(WebCore::RenderStyle::letterSpacing const): Deleted.
(WebCore::RenderStyle::setLetterSpacingWithoutUpdatingFontDescription): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::initialLetterSpacing): Deleted.
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialLetterSpacing):
(WebCore::RenderStyle::letterSpacing const):
(WebCore::RenderStyle::wordSpacing const):
* Source/WebCore/rendering/svg/RenderSVGInlineText.cpp:
(WebCore::RenderSVGInlineText::computeNewScaledFontForStyle):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextLengthOrNormal):
(WebCore::Style::BuilderConverter::convertSpacing): Deleted.
(WebCore::Style::BuilderConverter::convertWordSpacing): Deleted.
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInheritWordSpacing):
(WebCore::Style::BuilderCustom::applyInheritLetterSpacing):
(WebCore::Style::BuilderCustom::applyInitialLetterSpacing):
(WebCore::Style::BuilderCustom::applyValueLetterSpacing):
(WebCore::Style::applyLetterSpacing): Deleted.

Canonical link: https://commits.webkit.org/271357@main
